File: 396-rotate-function/rotate-function.cpp
class Solution {
public:
    int maxRotateFunction(vector<int>& nums) {
        int n = nums.size();
        int sum = 0;
        long long curr = 0;
        for (int i = 0; i < n; i++) {
            sum += nums[i];
            curr += i * nums[i];
        }
        long long ans = curr;
        // curr = curr + sum - n * nums[i];
        int j = n - 1;
        while (j > 0) {
            curr = curr + sum - (long long)n * nums[j];
            ans = max(curr, ans); j--;
        }
        return ans;
    }
};

File: 396-rotate-function/rotate-function.ts
function maxRotateFunction(nums: number[]): number {
    const n: number = nums.length;
    const sum = nums.reduce((total: number, value: number) => total + value, 0); // sum
    let curr = nums.reduce((total: number, value: number, i: number) => total + i * value, 0);// index wise sum
    let ans = curr;
    for (let i = n - 1; i > 0; i--) {
        curr = curr + sum - n * nums[i];
        ans = Math.max(ans, curr);
    }
    return ans;
};